### PR TITLE
[assets] added ability to manage cloudinary assets

### PR DESCRIPTION
**Pull request by OmerGronich**
1. add / edit storage with kind: cloudinary.
2. upload / remove / update / get list of assets from cloudinary API.

### CODE REVIEW

**Comment by OmerGronich** [comment]
@davidmeirlevy 

A few questions -

1. The maximum page size for a single request is 500 and the minimum is 10 in cloudinary. What is the expected response from get list of assets, do you want the full list, the minimum or 500? some number in the middle? let me know

23. The loadFiles method is expected to return an object in a specific shape, with a property called updated however Cloudinary does not provide updated_at field for their assets, only created_at. Do you want me to skip this property?

**Comment by davidmeirlevy** [comment]
@OmerGronich 
1. the expectation is to get the list of a specific path.
2. skip or set updated_at with the created_at (probably the same meaning... )

File: apps/assets/server/models/cloudinary.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        secure: true
+      });
+    });
+  }
+
+  list(prefix = '/') {
+
+    if (prefix === '/') {
+      // load all resources
+      return cloudinary.api.resources();
```
Comment (davidmeirlevy) [inline apps/assets/server/models/cloudinary.js:22]: `/` should return only the resources on the main "folder". not all.

File: apps/assets/server/models/cloudinary.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+class Cloudinary {
+  constructor(storage) {
+    this.name = storage.name;
+    this.ready = getSecret(storage.tenant, storage.authentication).then((decrypted) => {
+      const auth = decrypted.value;
+      cloudinary.config({
+        cloud_name: storage.metadata.bucketName,
+        api_key: auth.accessKeyId,
+        api_secret: auth.secretAccessKey,
```
Comment (davidmeirlevy) [inline apps/assets/server/models/cloudinary.js:12]: use Cloudinary's names for the variables from the secrets. e.g.: apiKey, apiSecret.
prefer even to use the same name from their UI interface at Cloudinary's dashboard.

File: apps/assets/server/services/cloudinary.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+async function uploadFile(storage, { identifier, file, extension, prefix }) {
+  const cloudinary = new Cloudinary(storage);
+  const filename = generateIdentifier(prefix, extension);
+  const folder = path.join(storage.metadata.basePath, identifier);
+  const parser = new DatauriParser();
+  const { content } = parser.format(filename, file);
+
+  const options = Object.assign({
```
Comment (davidmeirlevy) [inline apps/assets/server/services/cloudinary.js:44]: prefer not using Object.assign(). use regular spread.

File: apps/assets/server/models/cloudinary.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  remove(publicId) {
+    return cloudinary.api.delete_resources([publicId]);
+  }
+
+  rename(oldIdentifier, newFileName) {
+    return cloudinary.uploader.rename(oldIdentifier, newFileName);
+  }
+}
```
Comment (davidmeirlevy) [inline apps/assets/server/models/cloudinary.js:44]: you're missing a "destroy" method to kill the connection.

File: apps/assets/server/services/cloudinary.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  try {
+    await cloudinary.ready;
+    await cloudinary.rename(oldIdentifier, newFileName);
+  } catch (e) {
+    throw new Error(e.message || 'failed to remove asset: ' + fullPath);
+  }
+
+  return { success: true };
+}
+
```
Comment (davidmeirlevy) [inline apps/assets/server/services/cloudinary.js:88]: you're missing to destroy the connection at the end of every method.

File: apps/assets/server/services/cloudinary.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+async function renameFile(storage, oldIdentifier, newFileName) {
+  const cloudinary = new Cloudinary(storage);
+  const fullPath = path.join(storage.metadata.basePath || '/', oldIdentifier);
+
+  try {
+    await cloudinary.ready;
+    await cloudinary.rename(oldIdentifier, newFileName);
```
Comment (davidmeirlevy) [inline apps/assets/server/services/cloudinary.js:81]: the identifier is the full path to the file, and the new name is probably the specific file name.
e.g.:
```
oldIdentifier = /my/path/to/logo.png
newFileName = old-logo.png
```

**Review by davidmeirlevy — COMMENTED** [review]
great work.
only a few little issues.
also - try to revert the commits to all the package-locks files. they were probably changes because of linter issues and not because of actual dependencies changes (except the assets service)

File: apps/assets/server/models/cloudinary.js (hunk 1 above)
Comment (OmerGronich) [reply apps/assets/server/models/cloudinary.js]: fixed

File: apps/assets/server/models/cloudinary.js (hunk 2 above)
Comment (OmerGronich) [reply apps/assets/server/models/cloudinary.js]: done

File: apps/assets/server/services/cloudinary.js (hunk 3 above)
Comment (OmerGronich) [reply apps/assets/server/services/cloudinary.js]: done

File: apps/assets/server/models/cloudinary.js (hunk 4 above)
Comment (OmerGronich) [reply apps/assets/server/models/cloudinary.js:37]: I don't think there is a connection to kill. All the calls are stateless. I can't find in the cloudinary SDK any function the kills a connection as well.

File: apps/assets/server/services/cloudinary.js (hunk 5 above)
Comment (OmerGronich) [reply apps/assets/server/services/cloudinary.js:84]: ^

File: apps/assets/server/services/cloudinary.js (hunk 6 above)
Comment (OmerGronich) [reply apps/assets/server/services/cloudinary.js]: fixed

File: apps/assets/server/models/storage.js [hunk 7]
```diff
@@ -19,7 +19,7 @@ const StorageSchema = new mongoose.Schema({
   metadata: {
     publicUrl: {
       type: String,
-      required: true
+      required: isPublicUrlRequired
```
Comment (davidmeirlevy) [inline apps/assets/server/models/storage.js:22]: The public URL should always be required.
The scenario is when the given URL from the provider is different than the actual URL.

For example, assume I have my own domain, the I pass request from "images.mydomain.com" to my cloudinary account, using an "A RECORD" or "CNAME RECORD" to the cloudinary bucket, which can be something like "my-account-name.something.cloudinary.io".

From cloudinary's perspective, the base URL is their's, but from my own website, I want users to get images from my private domain (e.g. images.mydomain.com).

So this field is here to define my storage base URL, so when this service returns the metadata of an asset - its URL must be from the defined domain, and not cloudinary's.

**Review by davidmeirlevy — COMMENTED** [review]
Hi!

Last change regarding the storage public URL, as described.
But that's all ;)

File: apps/assets/server/models/storage.js (hunk 7 above)
Comment (OmerGronich) [reply apps/assets/server/models/storage.js]: Oh ok. Got it. Thought this field was something else.

Reverted the StorageSchema publicUrl field to be required again

**Review by davidmeirlevy — APPROVED** [review]

**Comment by davidmeirlevy** [comment]
Looks good. Have you tested it?

**Comment by OmerGronich** [comment]
> Looks good. Have you tested it?

Yup, I can show you when you have a few minutes to spare.

**Pull request merged**